### PR TITLE
[Feat] MKD

### PR DESCRIPTION
**Pull request by manyids2**
### Description
A few sentences describing the changes proposed in this pull request.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Review by ducha-aiki — APPROVED** [review]
LGTM

**Comment by ducha-aiki** [comment]
@edgarriba Could you please take a quick look? For me, everything looks good, just want to make sure I didn't miss anything

File: kornia/feature/mkd.py [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+        elif dtype == 'cart':
+            self.odims = cart_emb.odims
+        elif dtype == 'polar':
+            self.odims = polar_emb.odims
+
+        # Compute true reduce_dims.
+        self.reduce_dims: int = min(reduce_dims, self.odims)
+
+        # Load supervised(lw)/unsupervised(pca) model trained on training_set.
+        if self.whitening is not None:
```
Comment (edgarriba) [inline kornia/feature/mkd.py:521]: how can this be None if you define it as a `str` ?

File: kornia/feature/mkd.py [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+                urls[self.dtype], map_location=lambda storage, loc: storage
+            )
+            whitening_model = whitening_models[training_set]
+            self.whitening_layer = Whitening(whitening,
+                                             whitening_model,
+                                             in_dims=self.odims,
+                                             reduce_dims=self.reduce_dims)
+            self.odims = self.reduce_dims
+
+    def forward(self, patches: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/feature/mkd.py:532]: check input

File: kornia/feature/mkd.py [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+        # Extract gradients.
+        g = self.smoothing(patches)
+        g = self.gradients(g)
```
Comment (edgarriba) [inline kornia/feature/mkd.py:535]: might be easier to have it in a Sequential ?

File: kornia/feature/mkd.py [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+                                         reduce_dims=128)
+        >>> desc = mkd(patches) # 23x128
+    """
+
+    def __init__(self,
+                 patch_size: int = 32,
+                 dtype: str = 'concat',
+                 whitening: str = 'pcawt',
+                 training_set: str = 'liberty',
+                 reduce_dims: int = 128) -> None:
```
Comment (edgarriba) [inline kornia/feature/mkd.py:477]: `output_dims` might be more meaningful

File: kornia/feature/mkd.py [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+                                             in_dims=self.odims,
+                                             reduce_dims=self.reduce_dims)
+            self.odims = self.reduce_dims
+
+    def forward(self, patches: torch.Tensor) -> torch.Tensor:
+        # Extract gradients.
+        g = self.smoothing(patches)
+        g = self.gradients(g)
+
+        # Extract polar and/or cart features.
```
Comment (edgarriba) [inline kornia/feature/mkd.py:537]: I don't get how is this used. Can the user pass both polar/cart and/or concat? Can this better be resolved inside the constructor. To much ifelse branching here and below

File: kornia/feature/mkd.py [hunk 6]
```diff
@@ -0,0 +1,609 @@
...
+
+def load_whitening_model(dtype: str, training_set: str) -> Dict:
+    whitening_models = torch.hub.load_state_dict_from_url(
+        urls[dtype], map_location=lambda storage, loc: storage
+    )
+    whitening_model = whitening_models[training_set]
+    return whitening_model
+
+
+class SimpleKD(nn.Module):
```
Comment (edgarriba) [inline kornia/feature/mkd.py:577]: is this exposed to the highlevel api too ?

File: test/feature/test_mkd.py [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+
+
+@pytest.mark.parametrize("d1,d2",
+                         [(1, 1), (1, 2), (2, 1), (5, 6)])
+def test_get_kron_order(d1, d2):
+    out = get_kron_order(d1, d2)
+    assert out.shape == (d1 * d2, 2)
+
+
+class TestMKDGradients:
```
Comment (edgarriba) [inline test/feature/test_mkd.py:25]: tests are great! you missing `dtype` in each test. BTW, have you tried to run in GPU ?

Comment (edgarriba) [reply test/feature/test_mkd.py:25]: optional but would really help to implement as a class. Check: https://github.com/kornia/kornia/blob/master/test/enhance/test_normalize.py#L223

Comment (edgarriba) [reply test/feature/test_mkd.py:25]: can you also try test with JIT ?

**Review by edgarriba — CHANGES_REQUESTED** [review]
Great PR ! Lots of new useful functionalities. Check my comments and feel free to discuss anything. Don't forget to add the functionalities to the docs and build `make build-docs`

File: kornia/__init__.py [hunk 8]
```diff
@@ -67,7 +67,8 @@
     harris_response,
     hessian_response,
     gftt_response,
-    SIFTDescriptor
+    SIFTDescriptor,
+    MKD
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/__init__.py:71]: Done

File: kornia/geometry/conversions.py [hunk 9]
```diff
@@ -69,6 +72,54 @@ def deg2rad(tensor: torch.Tensor) -> torch.Tensor:
...
 
 
+def pol2cart(rho: torch.Tensor, phi: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    r"""Function that converts polar coordinates to cartesian coordinates.
+
+    Args:
+        rho (torch.Tensor): Tensor of arbitrary shape.
+        phi (torch.Tensor): Tensor of same arbitrary shape.
+
+    Returns:
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/geometry/conversions.py:82]: not sure which part you are talking about?

File: kornia/geometry/conversions.py [hunk 10]
```diff
@@ -69,6 +72,54 @@ def deg2rad(tensor: torch.Tensor) -> torch.Tensor:
...
+
+
+def cart2pol(x: torch.Tensor, y: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    r"""Function that converts cartesian coordinates to polar coordinates.
+
+    Args:
+        rho (torch.Tensor): Tensor of arbitrary shape.
+        phi (torch.Tensor): Tensor of same arbitrary shape.
+
+    Returns:
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/geometry/conversions.py:106]: not sure which part you are talking about?

File: test/geometry/test_conversions.py [hunk 11]
```diff
@@ -368,6 +368,46 @@ def test_deg2rad(batch_shape, device):
                      raise_exception=True)
 
 
+@pytest.mark.parametrize("batch_shape", [
+    (2, 3), (1, 2, 3), (2, 3, 3), (5, 5, 3), ])
+def test_pol2cart(batch_shape, device):
```
[the remark this answers is not in the document]
Comment (manyids2) [reply test/geometry/test_conversions.py:373]: I used test_sift.py as template, where classes have test classes, and functions have test functions.

File: kornia/feature/mkd.py [hunk 12]
```diff
@@ -0,0 +1,609 @@
+from typing import Union, Tuple, Dict, List
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:1]: Above or below?
siftdesc.py does not have newlines.

File: kornia/feature/mkd.py (hunk 3 above)
Comment (manyids2) [reply kornia/feature/mkd.py:535]: Please see comment below about branching.

File: kornia/feature/mkd.py (hunk 4 above)
Comment (manyids2) [reply kornia/feature/mkd.py:477]: done

File: kornia/feature/mkd.py (hunk 5 above)
Comment (manyids2) [reply kornia/feature/mkd.py:537]: Basially, for polar and cart, the model looks like:
`input -> smooth -> grads -> Position embedding -> Whiten`
but in the concat case,
```
input -> smooth -> grads -> pos1 -\
                           - pos2 -> concat -> Whiten
```
So it breaks the sequential chain.
I can try to refactor to remove branching by using dicts instead.

File: kornia/feature/mkd.py (hunk 6 above)
Comment (manyids2) [reply kornia/feature/mkd.py:577]: No, it was purely an example to illustrate that without 'concat', the whole thing can be written as a sequence.

File: test/feature/test_mkd.py (hunk 7 above)
Comment (manyids2) [reply test/feature/test_mkd.py:25]: GPU - i can see that there is a conftest.py file, that adds device as an argument, so i think it is running GPU. 
I am not familiar with JIT, so haven't coded any tests for it.

File: kornia/feature/mkd.py [hunk 13]
```diff
@@ -0,0 +1,609 @@
...
+COEFFS_N3_K8: List[float] = [0.14343168, 0.268285, 0.21979234, 0.15838885]
+COEFFS: Dict[str, List[float]] = {'xy': COEFFS_N1_K1,
+                                  'rhophi': COEFFS_N2_K8,
+                                  'theta': COEFFS_N3_K8}
+
+urls: Dict[str, str] = {k: f'https://github.com/manyids2/mkd_pytorch/raw/master/mkd_pytorch/mkd-{k}-64.pth'
+                        for k in ['cart', 'polar', 'concat']}
+
+
+def get_grid_dict(patch_size: int = 32) -> Dict[str, torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:25]: I tried it, but pytest is no longer able to import it, same for spatial_kernel_embedding.

File: kornia/feature/mkd.py [hunk 14]
```diff
@@ -0,0 +1,609 @@
...
+    """Gets order for doing kronecker product. """
+    kron_order = torch.zeros([d1 * d2, 2], dtype=torch.int64)
+    for i in range(d1):
+        for j in range(d2):
+            kron_order[i * d2 + j, 0] = i
+            kron_order[i * d2 + j, 1] = j
+    return kron_order
+
+
+class MKDGradients(nn.Module):
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:47]: Again, not sure what the change must be.

File: kornia/feature/mkd.py [hunk 15]
```diff
@@ -0,0 +1,609 @@
...
+    def __init__(self) -> None:
+        super().__init__()
+        self.eps = 1e-8
+
+        # Modify 'diff' gradient.
+        grad_fn = SpatialGradient(mode='diff', order=1, normalized=False)
+        grad_fn.kernel = -1 * grad_fn.kernel
+        self.grad = grad_fn
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:76]: done

File: kornia/feature/mkd.py [hunk 16]
```diff
@@ -0,0 +1,609 @@
...
+        oris = torch.atan2(gy, gx)
+        y = torch.cat([mags, oris], dim=1)
+        return y
+
+    def __repr__(self) -> str:
+        return self.__class__.__name__
+
+
+class VonMisesKernel(nn.Module):
+    """
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:96]: same reply

File: kornia/feature/mkd.py [hunk 17]
```diff
@@ -0,0 +1,609 @@
...
+        >>> emb = vm(oris) # 23x7x32x32
+    """
+
+    def __init__(self,
+                 patch_size: int,
+                 coeffs: Union[list, tuple]) -> None:
+        super().__init__()
+
+        self.patch_size = patch_size
+        b_coeffs: torch.Tensor = torch.Tensor(coeffs).float()
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: done

File: kornia/feature/mkd.py [hunk 18]
```diff
@@ -0,0 +1,609 @@
...
+        self.register_buffer('coeffs', b_coeffs)
+
+        # Compute parameters.
+        n: int = len(coeffs) - 1
+        self.n: int = n
+        self.d: int = 2 * n + 1
+
+        # Precompute helper variables.
+        emb0 = torch.ones([1, 1, patch_size, patch_size]).float()
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: done

File: kornia/feature/mkd.py [hunk 19]
```diff
@@ -0,0 +1,609 @@
...
+    """
+
+    def __init__(self,
+                 patch_size: int,
+                 coeffs: Union[list, tuple]) -> None:
+        super().__init__()
+
+        self.patch_size = patch_size
+        b_coeffs: torch.Tensor = torch.Tensor(coeffs).float()
+        self.register_buffer('coeffs', b_coeffs)
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:124]: or else model.to(device) was not working as intended.

File: kornia/feature/mkd.py [hunk 20]
```diff
@@ -0,0 +1,609 @@
...
+        weights[:n + 1] = torch.sqrt(b_coeffs)
+        weights[n + 1:] = torch.sqrt(b_coeffs[1:])
+        weights = weights.reshape(-1, 1, 1).float()
+        self.register_buffer('emb0', emb0)
+        self.register_buffer('frange', frange)
+        self.register_buffer('weights', weights)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        emb0 = self.emb0.repeat(x.size(0), 1, 1, 1)  # type: ignore
+        frange = self.frange * x
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: done

File: kornia/feature/mkd.py [hunk 21]
```diff
@@ -0,0 +1,609 @@
...
+        # Theta kernel for gradients.
+        self.kernel = VonMisesKernel(patch_size=patch_size,
+                                     coeffs=COEFFS['theta'])
+
+        # Relative gradients.
+        kgrid = create_meshgrid(height=patch_size,
+                                width=patch_size,
+                                normalized_coordinates=True)
+        _, phi = cart2pol(kgrid[:, :, :, 0], kgrid[:, :, :, 1])
+        self.register_buffer('phi', phi.float())
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: done

File: kornia/feature/mkd.py [hunk 22]
```diff
@@ -0,0 +1,609 @@
...
+        frange = frange.reshape(-1, 1, 1).float()
+        weights = torch.zeros([2 * n + 1]).float()
+        weights[:n + 1] = torch.sqrt(b_coeffs)
+        weights[n + 1:] = torch.sqrt(b_coeffs[1:])
+        weights = weights.reshape(-1, 1, 1).float()
+        self.register_buffer('emb0', emb0)
+        self.register_buffer('frange', frange)
+        self.register_buffer('weights', weights)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:143]: done

File: kornia/feature/mkd.py [hunk 23]
```diff
@@ -0,0 +1,609 @@
...
+        emb0 = self.emb0.repeat(x.size(0), 1, 1, 1)  # type: ignore
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: size is not callable error.

File: kornia/feature/mkd.py [hunk 24]
```diff
@@ -0,0 +1,609 @@
...
+    r"""
+    Module, which computes gradients of given patches,
+    stacked as [magnitudes, orientations].
+    Given gradients $g_x$, $g_y$ with respect to $x$, $y$ respectively,
+      - $\mathbox{mags} = $\sqrt{g_x^2 + g_y^2 + eps}$
+      - $\mathbox{oris} = $\mbox{tan}^{-1}(\nicefrac{g_y}{g_x})$.
+    Args:
+        patch_size: (int) Input patch size in pixels (32 is default)
+    Returns:
+        Tensor: gradients of given patches
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:57]: done

File: kornia/feature/mkd.py [hunk 25]
```diff
@@ -0,0 +1,609 @@
...
+                                normalized_coordinates=True)
+        _, phi = cart2pol(kgrid[:, :, :, 0], kgrid[:, :, :, 1])
+        self.register_buffer('phi', phi.float())
+
+    def emb_mags(self, mags: torch.Tensor) -> torch.Tensor:
+        """Embed square roots of magnitudes with eps for numerical reasons. """
+        mags = torch.sqrt(mags + self.eps)
+        return mags
+
+    def forward(self, grads: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:209]: done

File: kornia/feature/mkd.py [hunk 26]
```diff
@@ -0,0 +1,609 @@
...
+        y = self.kernel(oris) * self.emb_mags(mags)
+        return y
+
+    def __repr__(self) -> str:
+        return self.__class__.__name__ +\
+            '(' + 'patch_size=' + str(self.patch_size) +\
+            ', ' + 'relative=' + str(self.relative) + ')'
+
+
+def spatial_kernel_embedding(dtype, grids: dict) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: pytest does not seem to pickup functions that have underscore

File: kornia/feature/mkd.py [hunk 27]
```diff
@@ -0,0 +1,609 @@
...
+        >>> ese = kornia.feature.mkd.ExplicitSpacialEncoding(dtype='polar',
+                                                             fmap_size=32,
+                                                             in_dims=7,
+                                                             do_gmask=True,
+                                                             do_l2=True)
+        >>> desc = ese(emb_ori) # 23x175x32x32
+    """
+
+    def __init__(self,
+                 dtype: str = 'polar',
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: done

File: kornia/feature/mkd.py [hunk 28]
```diff
@@ -0,0 +1,609 @@
...
+        gmask = torch.exp(-1 * norm_rho**2 / sigma**2)
+        return gmask
+
+    def init_kron(self) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Initialize helper variables to calculate kronecker. """
+        kron = get_kron_order(self.in_dims, self.d_emb)
+        emb2 = torch.index_select(self.emb, 1, kron[:, 1])  # type: ignore
+        return emb2, kron[:, 0]
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:337]: done

File: kornia/feature/mkd.py [hunk 29]
```diff
@@ -0,0 +1,609 @@
...
+            ', ' + 'in_dims=' + str(self.in_dims) +\
+            ', ' + 'out_dims=' + str(self.out_dims) +\
+            ', ' + 'do_gmask=' + str(self.do_gmask) +\
+            ', ' + 'do_l2=' + str(self.do_l2) + ')'
+
+
+class Whitening(nn.Module):
+    """
+    Module, performs supervised or unsupervised whitening as described in
+    [Understanding and Improving Kernel Local Descriptors](https://arxiv.org/abs/1811.11147) .
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: done

File: kornia/feature/mkd.py [hunk 30]
```diff
@@ -0,0 +1,609 @@
...
+        self.keval = keval
+        self.t = t
+        self.pval = 1.0
+
+        # Compute true reduce_dims.
+        reduce_dims = min(reduce_dims, in_dims)
+        self.reduce_dims = reduce_dims
+
+        # Initialize identity transform.
+        self.mean = nn.Parameter(torch.zeros(in_dims).float(),
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: done

File: kornia/feature/mkd.py [hunk 31]
```diff
@@ -0,0 +1,609 @@
...
+        return self.__class__.__name__ +\
+            '(' + 'dtype=' + str(self.dtype) +\
+            ', ' + 'fmap_size=' + str(self.fmap_size) +\
+            ', ' + 'in_dims=' + str(self.in_dims) +\
+            ', ' + 'out_dims=' + str(self.out_dims) +\
+            ', ' + 'do_gmask=' + str(self.do_gmask) +\
+            ', ' + 'do_l2=' + str(self.do_l2) + ')'
+
+
+class Whitening(nn.Module):
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:361]: Will check and let you know.

File: kornia/feature/mkd.py [hunk 32]
```diff
@@ -0,0 +1,609 @@
...
+        alpha = self.evals[self.keval]
+        evals = ((1 - alpha) * self.evals) + alpha
+        self.evecs.data = self.evecs @ torch.diag(torch.pow(evals, -0.5))
+
+    def _modify_pcawt(self) -> None:
+        """ Attenuation for eigenvalues."""
+        m = -0.5 * self.t
+        self.evecs.data = self.evecs @ torch.diag(torch.pow(self.evals, m))
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py:454]: done

File: kornia/feature/mkd.py [hunk 33]
```diff
@@ -0,0 +1,609 @@
...
+        return self.__class__.__name__ +\
+            '(' + 'xform=' + str(self.xform) +\
+            ', ' + 'in_dims=' + str(self.in_dims) +\
+            ', ' + 'reduce_dims=' + str(self.reduce_dims) + ')'
+
+
+class MKD(nn.Module):
+    """
+    Module that computes Multiple Kernel local descriptors as described in
+    [Understanding and Improving Kernel Local Descriptors](https://arxiv.org/abs/1811.11147) .
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: done

File: kornia/feature/mkd.py [hunk 34]
```diff
@@ -0,0 +1,609 @@
...
+                                         dtype='concat',
+                                         whitening='pcawt',
+                                         training_set='liberty',
+                                         reduce_dims=128)
+        >>> desc = mkd(patches) # 23x128
+    """
+
+    def __init__(self,
+                 patch_size: int = 32,
+                 dtype: str = 'concat',
```
[the remark this answers is not in the document]
Comment (manyids2) [reply kornia/feature/mkd.py]: done

File: kornia/feature/mkd.py (hunk 1 above)
Comment (manyids2) [reply kornia/feature/mkd.py:553]: Maybe is should use the string 'none' ?

File: kornia/feature/mkd.py (hunk 2 above)
Comment (manyids2) [reply kornia/feature/mkd.py:564]: done

File: kornia/geometry/conversions.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/geometry/conversions.py:82]: No dashes in your code, I guess. Try to do `make build-docs` and check if the docs are displayed ok

File: test/geometry/test_conversions.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply test/geometry/test_conversions.py:373]: we are refactoring tests to have full coverage. Sift test might need to be updated too at some point. It's just a suggestion that will help us to have less work later.

File: kornia/feature/mkd.py (hunk 5 above)
Comment (edgarriba) [reply kornia/feature/mkd.py:575]: great

File: kornia/feature/mkd.py (hunk 6 above)
Comment (edgarriba) [reply kornia/feature/mkd.py:615]: ok then I suggest that everything that won't go to the high level API goes with underscore in front. This includes classes too

File: test/feature/test_mkd.py (hunk 7 above)
Comment (edgarriba) [reply test/feature/test_mkd.py:25]: check this: https://github.com/kornia/kornia/blob/master/test/enhance/test_normalize.py#L273
if everything it's well done, should be plug and play.

**Comment by ducha-aiki** [comment]
@manyids you can click "resolve" on those Edgar comments, which are "done", so we can revisit it again and merge :)

**Comment by edgarriba** [comment]
@manyids2 can you verify that you rebased correctly ? it seems that is taking somehow the old circleCI setup file which not includes TPU testing :)

**Comment by manyids** [comment]
I did not use rebase, I pulled and checked out feat/MKD branch... How
should I proceed?

**Comment by edgarriba** [comment]
@manyids2 to rebase properly try `git pull --rebase upstream master`.

**Comment by edgarriba** [comment]
@fthorde the rebase didn't work well. @ducha-aiki can help with this task? we will release next week and we might want this PR in

**Comment by edgarriba** [comment]
merged in https://github.com/kornia/kornia/pull/841

**Pull request closed without merging**